performance/runner/src/fs.rs

```diff
@@ -0,0 +1,349 @@
+        .concat();
+
+    Ok(results)
+}
+
+// reads directory names under this directory and converts their names to sem-ver versions.
+// returns the latest version of the set.
+//
+// this is used to identify which baseline version we should be targeting to compare samples against.
+pub fn latest_version_from(dir: &dyn AsRef<Path>) -> Result<Version, RunnerError> {
+    let versions = all_dirs_in(dir)?
+        .into_iter()
+        // this line is a little opaque but it's just converting OsStr -> String with options along the way.
+        .filter_map(|d| {
+            d.file_name()
+                .and_then(|fname| fname.to_str().map(|x| x.to_string()))
+        })
+        .map(|fname| Version::from_str(&fname))
+        .collect::<Result<Vec<Version>, RunnerError>>()?;
+
+    versions
+        .into_iter()
+        .reduce(cmp::max)
+        .ok_or_else(|| RunnerError::NoVersionedBaselineData(dir.as_ref().to_path_buf()))
+}
+
+fn all_dirs_in(dir: &dyn AsRef<Path>) -> Result<Vec<PathBuf>, IOError> {
+    Ok(fs::read_dir(dir)
+        .or_else(|e| Err(IOError::ReadErr(dir.as_ref().to_path_buf(), Some(e))))?
+        .collect::<Result<Vec<DirEntry>, io::Error>>()
+        .or_else(|e| Err(IOError::ReadIterErr(dir.as_ref().to_path_buf(), Some(e))))?
+        .into_iter()
+        .filter_map(|d| {
+            let path = d.path();
+            if path.is_dir() {
+                Some(path)
+            } else {
+                None
+            }
+        })
+        .collect())
+}
+
+// TODO can we call hyperfine as a rust library?
+// https://crates.io/crates/hyperfine/1.13.0
+fn run_hyperfine(
+    run_dir: &dyn AsRef<Path>,
+    command: &str,
+    prep: &str,
+    runs: i32,
+    output_file: &dyn AsRef<Path>,
+) -> Result<ExitStatus, IOError> {
+    Command::new("hyperfine")
+        .current_dir(run_dir)
+        // warms filesystem caches by running the command first without counting it.
+        // alternatively we could clear them before each run
+        .arg("--warmup")
+        .arg("1")
+        // --min-runs defaults to 10
+        .arg("--min-runs")
+        .arg(runs.to_string())
+        .arg("--max-runs")
+        .arg(runs.to_string())
+        .arg("--prepare")
+        .arg(prep)
+        .arg(command)
+        .arg("--export-json")
+        .arg(output_file.as_ref())
+        // this prevents hyperfine from capturing dbt's output.
+        // Noisy, but good for debugging when tests fail.
+        .arg("--show-output")
+        .status() // use spawn() here instead for more information
+        .or_else(|e| Err(IOError::CommandErr(Some(e))))
+}
+
+// Attempt to delete the directory and its contents. If it doesn't exist we'll just recreate it anyway.
+fn clear_dir(dir: &dyn AsRef<Path>) -> Result<(), io::Error> {
+    match fs::remove_dir_all(dir) {
+        // whether it existed or not, create the directory.
+        _ => fs::create_dir(dir),
```

Comment:
what do you mean by parent directory? this should just be deleting a directory and all of its contents (child directories included) then recreating it empty.

Reply:
ah ya, poor wording on my part

`create_dir("parent/child")` -> `mkdir parent/child` will fail if parent dir doesn't exist
`create_dir_all("parent/child")` -> `mkdir -p parent/child`

not an issue here if `dir` here is just 1 level ever

Reply:
no preference on match vs combinator

Reply:
right. if users give me a dud path my only option is to give a clear error. which I think is what you're pointing out here: a bad path doesn't have a great error story here. I'll improve that.

Reply:
1- the `and_then` syntax isn't semantically equivalent. oops.
2- I changed it to create any missing parents as well instead of an error. equivalent of `mkdir -p`